### PR TITLE
fix(TradeInput): max ETH Trade bug covering gas

### LINKED ISSUE #1280: Trademax error when attempting to make a max ETH trade in the trade widget.
### Overview

getting an error when attempting to make a max ETH trade in the trade widget. keep key and native have the error, "An error occurred getting a quote for your trade." MetaMask and Portis are getting the error , "Not enough ETH to cover gas."

### References and additional details

1. go to app.shapeshift or https://2f6f7ed6.web-29e.pages.dev/
2. log in with your mm, kk, portis, or native wallet.
3. go to the trade widget toward the upper left hand side of the screen.
4. with ETH selected as the input asset, click the max send button.
5. observer either, "An error occurred getting a quote for your trade." for kk and native, or "Not enough ETH to cover gas." for MM and Portis.

### Acceptance Criteria

the trade widget should allow for a max ETH trade and calculate your balance and gas properly 

### Need By Date

_No response_

### Screenshots/Mockups

Screenshot 2022-03-24 095925
Screenshot 2022-03-24 100033

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Sponsor / Stakeholder

_No response_

### Bounty Hunters

- [Join our discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info
- Please ***do not*** start work on this issue until you are approved in Gitcoin.

### PR DESCRIPTION
**Pull request by brymut**
## Description

Fixed error affecting the Max option while trading ETH to include gas fees, moved gas estimation to be included before value is populated in input.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #1280 

## Risk

To risks that I'm aware of.

## Testing

(Tested using MetaMask on ETH Mainnet
1. Go to the trade widget toward the upper left hand side of the screen.
2. With ETH selected as the input asset, click the max send button.

## Screenshots (if applicable)
https://user-images.githubusercontent.com/14836056/164548521-7d553bb8-f214-4071-a81b-8f086de1d9fa.mov

### CODE REVIEW

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 1]
```diff
@@ -117,13 +121,22 @@ export const useSwapper = () => {
 
     if (!minimumQuote) return
 
+    // to account for gas fees when trading Max ETH
+    const gasDeduction =
```
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:125]: We want to always account for gas fees, not only when sending max.
See the review comment.

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Hi @brymut, thanks for the contribution!

Tested this one out locally and this breaks the user-input amount, I can write down any amount up to the exact same amount I have in my wallet, without being blocked from continuing all the way to broadcast.

It will then produce an error toast in native/an error in Metamask, since the (full balance + gas) amount will obviously be greater than the full balance.

File: src/components/Trade/hooks/useSwapper/useSwapper.ts (hunk 1 above)
Comment (brymut) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:125]: Right, sorry, not sure how I missed that. Will fix asap

Comment (brymut) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:125]: Hi @gomesalexandre, I think I have resolved the issue now, however I believe the `estimatedGasFees` value in the `TradeInput` component is higher than the actual estimated gas fee used to create the quote for the transaction so the Max button doesn't result in a very good actual Max and the check to see if there is enough ETH for fees is using the significantly higher gas estimate. I've attached a screenrecording with the `estimatedGasFees` var being logged in the console

https://user-images.githubusercontent.com/14836056/164868363-b437fa7b-872e-4841-9c55-18cd22336108.mov

**Comment by gomesalexandre** [comment]
@brymut can you change your PR title to `fix(TradeInput): max ETH Trade bug covering gas` and push this empty commit: `git commit --allow-empty -m "chore: trigger CI"` after the title is changed, to make lint CI happy?

**Comment by brymut** [comment]
Hi @gomesalexandre, could you please have a look?

**Comment by gomesalexandre** [comment]
> Hi @gomesalexandre, could you please have a look?

Hi @brymut , still waiting for test funds to test this one, asked earlier this weekend - will take a look as soon as I receive them.

File: src/components/Trade/hooks/useSwapper/useSwapper.ts (hunk 1 above)
Comment (gomesalexandre) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:125]: > Hi @gomesalexandre, I think I have resolved the issue now, however I believe the `estimatedGasFees` value in the `TradeInput` component is higher than the actual estimated gas fee used to create the quote for the transaction so the Max button doesn't result in a very good actual Max and the check to see if there is enough ETH for fees is using the significantly higher gas estimate. I've attached a screenrecording with the `estimatedGasFees` var being logged in the console
> Screen.Recording.2022-04-23.at.5.31.39.AM.mov

@brymut tried your implementation and  I'm consistently getting "Not enough ETH to cover gas" error on the first try. It only works when clicking "Max" a second time.

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
See comment about the regression

File: src/components/Trade/hooks/useSwapper/useSwapper.ts (hunk 1 above)
Comment (brymut) [reply src/components/Trade/hooks/useSwapper/useSwapper.ts:127]: I think I have addressed the issue @gomesalexandre, problem was that estimatedGasFees was still undefined and wasn't being factored in, I have added that uses the quote fees to address this. 

https://user-images.githubusercontent.com/14836056/165241979-921f9372-ee7d-49e1-93aa-d3f184f7d299.mov

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 2]
```diff
@@ -116,20 +120,28 @@ export const useSwapper = () => {
...
 
     if (!minimumQuote) return
+    //use gas estimate from quote if estimatedGasFees var is not populated yet
+    const quoteEstimatedGasFees = fromBaseUnit(minimumQuote?.feeData?.fee || 0, feeAsset.precision)
+
+    // to account for gas fees when trading Max ETH
+    const gasDeduction =
+      sellAsset && feeAsset && feeAsset.caip19 === sellAsset.currency.caip19
+        ? bnOrZero(estimatedGasFees || quoteEstimatedGasFees)
+        : bnOrZero(0)
```
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:130]: I know we do `bnOrZero(0)` a lot around the codebase but that's wrong, it doesn't make sense to fallback to zero if you're already passing zero! What about `bn(0)` ? Also imported from `src/lib/bignumber/bignumber.ts`

**Review by gomesalexandre — APPROVED** [review]
LGTM now: I've tried to break it many times running it locally and I couldn't.

I'm not entirely sure if using the gas estimate from the quote while `estimatedGasFees` is the way to go (for accuracy reasons), but as far as the issue is concerned, that solves it. Well done @brymut !

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 3]
```diff
@@ -116,20 +120,28 @@ export const useSwapper = () => {
...
 
     const sendMaxAmount = await swapper.getSendMaxAmount({
       wallet,
       quote: minimumQuote,
       sellAssetAccountId: '0', // TODO: remove hard coded accountId when multiple accounts are implemented
     })
-
-    const formattedMaxAmount = fromBaseUnit(sendMaxAmount, sellAsset.currency.precision)
+    const formattedsendMaxAmount = fromBaseUnit(sendMaxAmount, sellAsset.currency.precision)
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:137]: Rename const to `formattedSendMaxAmount`

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 4]
```diff
@@ -116,20 +120,28 @@ export const useSwapper = () => {
     })
 
     if (!minimumQuote) return
+    //use gas estimate from quote if estimatedGasFees var is not populated yet
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:123]:
```suggestion
    // use gas estimate from quote if estimatedGasFees var is not populated yet
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 5]
```diff
@@ -116,20 +120,28 @@ export const useSwapper = () => {
...
+    const quoteEstimatedGasFees = fromBaseUnit(minimumQuote?.feeData?.fee || 0, feeAsset.precision)
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:124]: `0` is a valid `fee`.

```suggestion
    const quoteEstimatedGasFees = fromBaseUnit(minimumQuote?.feeData?.fee ?? 0, feeAsset.precision)
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 6]
```diff
@@ -116,20 +120,28 @@ export const useSwapper = () => {
...
+
+    // to account for gas fees when trading Max ETH
+    const gasDeduction =
+      sellAsset && feeAsset && feeAsset.caip19 === sellAsset.currency.caip19
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:128]: Our types above say `feeAsset` and `sellAsset` are always defined. 
Either these checks are not needed, or the types are incorrect.

```suggestion
      feeAsset.caip19 === sellAsset.currency.caip19
```

File: src/components/Trade/hooks/useSwapper/useSwapper.ts [hunk 7]
```diff
@@ -116,20 +120,28 @@ export const useSwapper = () => {
...
+        ? bnOrZero(estimatedGasFees || quoteEstimatedGasFees)
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useSwapper/useSwapper.ts:129]: We still want to accept a result of `0` from `estimatedGasFees` as valid, yeah?

```suggestion
        ? bnOrZero(estimatedGasFees ?? quoteEstimatedGasFees)
```

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Looks good @brymut pending some stylistic changes and nullish coalescing fixes before we merge.

**Comment by brymut** [comment]
Suggested changes applied

**Review by 0xApotheosis — APPROVED** [review]
Thank you ser

**Comment by elmutt** [comment]
These changes are very zrx specific. We are trying to move towards supporting multiple swapers and need to be more careful about changes like this

**Pull request merged**